library/Icingadb/Widget/ItemList/ObjectList.php

```diff
@@ -0,0 +1,296 @@
+            $this->isActionList = true;
+        }
+
+        $this->hasDetailUrl = true;
+
+        $this->addAttributes(new Attributes(['data-icinga-detail-url' => $url]));
+
+        return $this;
+    }
+
+    /**
+     * Associate the given element with the given multi-selection filter
+     *
+     * @param BaseHtmlElement $element
+     * @param Filter\Rule $filter
+     *
+     * @return $this
+     */
+    protected function addMultiselectFilterAttribute(BaseHtmlElement $element, Filter\Rule $filter): self
+    {
+        $queryString = QueryString::render($filter);
+        if ($filter instanceof Filter\Chain) {
+            $queryString = '(' . $queryString . ')';
+        }
+
+        $element->addAttributes(new Attributes(['data-icinga-multiselect-filter' => $queryString]));
+
+        return $this;
+    }
+
+    /**
+     * Associate the given element with the given single-selection filter
+     *
+     * @param BaseHtmlElement $element
+     * @param Filter\Rule $filter
+     *
+     * @return $this
+     */
+    protected function addDetailFilterAttribute(BaseHtmlElement $element, Filter\Rule $filter): self
+    {
+        $element->addAttributes(new Attributes([
+            'data-action-item'          => true,
+            'data-icinga-detail-filter' => QueryString::render($filter)
+        ]));
+
+        return $this;
+    }
+
+    public function getItemLayout($item): ItemLayout
+    {
+        $layout = parent::getItemLayout($item);
+        if ($this->hasIconImages()) {
+            $layout->after(ItemLayout::VISUAL, 'icon-image');
+        }
+
+        return $layout;
+    }
+
+    /**
+     * @param object<UnreachableParent|DependencyNode|RedundancyGroup|Service|Host> $data
+     *
+     * @return ListItem
+     */
+    protected function createListItem(object $data)
+    {
+        if ($this->isActionList && $this->hasMultiselectUrl && ! $this->hasDetailUrl) {
+            throw new LogicException('Multiselect URL is set but no detail URL');
+        }
+
+        $multiSelectFilter = null;
+        switch (true) {
+            case $data instanceof RedundancyGroup || isset($data->redundancy_group_id):
+                $object = $data->redundancy_group ?? $data;
+                $detailFilter = Filter::equal('id', bin2hex($object->id));
+                break;
+            case $data instanceof Service || isset($data->service_id):
+                $object = $data->service ?? $data;
+
+                $multiSelectFilter = Filter::all(
+                    Filter::equal('service.name', $object->name),
+                    Filter::equal('host.name', $object->host->name)
+                );
+
+                $detailFilter = Filter::all(
+                    Filter::equal('name', $object->name),
+                    Filter::equal('host.name', $object->host->name)
+                );
+
+                break;
+            case $data instanceof Host || isset($data->host_id):
```

Comment:
Please use a separate case for instances of `DependencyNode`, otherwise the risk for false positives is too high

Reply:
e.g. you assume the existence of a `service` relation, because a `service_id` is set and that this holds a value that is an instance of `Service`